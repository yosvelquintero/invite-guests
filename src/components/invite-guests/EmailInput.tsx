import React from 'react';

interface IEmailInputProps {
  email: string;
  onEmailChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onAddEmail: () => void;
  isEmailDuplicated: boolean;
  isEmailError: boolean;
}

const EmailInput = ({ email, onEmailChange, onAddEmail, isEmailDuplicated, isEmailError }: IEmailInputProps) => {
  const isAddDisabled = isEmailDuplicated || isEmailError || !email;

  return (
    <div className='email-input-section'>
      <input type='email' value={email} onChange={onEmailChange} placeholder='example@email.com' />
      <button onClick={onAddEmail} disabled={isAddDisabled}>
        Add
      </button>
      {isEmailDuplicated && <p className='error-message'>The entered email already exists.</p>}
    </div>
  );
};

export default EmailInput;
