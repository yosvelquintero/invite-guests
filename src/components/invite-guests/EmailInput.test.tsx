import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EmailInput from './EmailInput';

interface IEmailInputProps {
  email: string;
  onEmailChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onAddEmail: () => void;
  isEmailDuplicated: boolean;
  isEmailError: boolean;
}

describe('EmailInput Component', () => {
  // Mock functions for props
  const mockOnEmailChange = jest.fn();
  const mockOnAddEmail = jest.fn();

  // Utility function to render the component with default or custom props
  const renderComponent = (props: Partial<IEmailInputProps> = {}) => {
    const defaultProps: IEmailInputProps = {
      email: '',
      onEmailChange: mockOnEmailChange,
      onAddEmail: mockOnAddEmail,
      isEmailDuplicated: false,
      isEmailError: false,
    };

    return render(<EmailInput {...defaultProps} {...props} />);
  };

  // Reset mocks before each test to avoid interference
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the input field with correct placeholder', () => {
    renderComponent();

    const inputElement = screen.getByPlaceholderText('example@email.com');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute('type', 'email');
  });

  it('renders the Add button', () => {
    renderComponent();

    const addButton = screen.getByRole('button', { name: /add/i });
    expect(addButton).toBeInTheDocument();
    expect(addButton).toBeDisabled();
  });

  it('disables the Add button when email is duplicated', () => {
    renderComponent({ isEmailDuplicated: true, email: 'test@example.com' });

    const addButton = screen.getByRole('button', { name: /add/i });
    expect(addButton).toBeDisabled();
  });

  it('disables the Add button when there is an email error', () => {
    renderComponent({ isEmailError: true, email: 'invalid-email' });

    const addButton = screen.getByRole('button', { name: /add/i });
    expect(addButton).toBeDisabled();
  });

  it('disables the Add button when email input is empty', () => {
    renderComponent({ email: '' });

    const addButton = screen.getByRole('button', { name: /add/i });
    expect(addButton).toBeDisabled();
  });

  it('enables the Add button when email is valid and not duplicated or errored', () => {
    renderComponent({ email: 'test@example.com' });

    const addButton = screen.getByRole('button', { name: /add/i });
    expect(addButton).toBeEnabled();
  });

  it('calls onEmailChange when the input value changes', () => {
    renderComponent();

    const inputElement = screen.getByPlaceholderText('example@email.com');
    fireEvent.change(inputElement, { target: { value: 'some@email.com' } });

    expect(mockOnEmailChange).toHaveBeenCalledTimes(1);
    expect(mockOnEmailChange).toHaveBeenCalledWith(expect.any(Object));
  });

  it('calls onAddEmail when the Add button is clicked', async () => {
    renderComponent({ email: 'test@example.com' });

    const addButton = screen.getByRole('button', { name: /add/i });
    fireEvent.click(addButton);

    expect(mockOnAddEmail).toHaveBeenCalledTimes(1);
  });

  it('does not call onAddEmail when the Add button is disabled and clicked', async () => {
    renderComponent({ isEmailDuplicated: true, email: 'test@example.com' });

    const addButton = screen.getByRole('button', { name: /add/i });
    expect(addButton).toBeDisabled();

    fireEvent.click(addButton);

    expect(mockOnAddEmail).not.toHaveBeenCalled();
  });

  it('displays the duplicate email error message when isEmailDuplicated is true', () => {
    renderComponent({ isEmailDuplicated: true });

    const errorMessage = screen.getByText('The entered email already exists.');
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveClass('error-message');
  });

  it('does not display the duplicate email error message when isEmailDuplicated is false', () => {
    renderComponent({ isEmailDuplicated: false });

    const errorMessage = screen.queryByText('The entered email already exists.');
    expect(errorMessage).not.toBeInTheDocument();
  });

  it('matches the snapshot when rendered with default props', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when rendered with duplicated email', () => {
    const { asFragment } = renderComponent({ isEmailDuplicated: true, email: 'test@example.com' });
    expect(asFragment()).toMatchSnapshot();
  });
});
