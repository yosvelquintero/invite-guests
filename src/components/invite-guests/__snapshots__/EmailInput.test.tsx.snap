// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailInput Component matches the snapshot when rendered with default props 1`] = `
<DocumentFragment>
  <div
    class="email-input-section"
  >
    <input
      placeholder="example@email.com"
      type="email"
      value=""
    />
    <button
      disabled=""
    >
      Add
    </button>
  </div>
</DocumentFragment>
`;

exports[`EmailInput Component matches the snapshot when rendered with duplicated email 1`] = `
<DocumentFragment>
  <div
    class="email-input-section"
  >
    <input
      placeholder="example@email.com"
      type="email"
      value="test@example.com"
    />
    <button
      disabled=""
    >
      Add
    </button>
    <p
      class="error-message"
    >
      The entered email already exists.
    </p>
  </div>
</DocumentFragment>
`;
